### PR TITLE
[proposal] Caching member cluster resources for Karmada

### LINKED ISSUE #1245: [Proposal] caching member cluster resources
**What would you like to be added**:

Caching member cluster resources

Scenarios:
- Global resource access/search (multi-cluster/cross-regional cluster)
- Global application view (multi-cluster application)
- Unified multi-cluster resource entry (for dashboard/cli)

**Why is this needed**:

- Accelerate (cross-regional) resource access
- High-frequency kubernetes resource access
- Aggregate multi-cluster resources

### PR DESCRIPTION
**Pull request by huntsman-li**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Caching member cluster resources for Karmada

**Which issue(s) this PR fixes**:
Fixes #1245 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Thanks @huntsman-li , I'll look at it ASAP.

cc @GitHubxsy

File: docs/proposals/caching/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+- Get the distribution of this application in different clusters (resource aggregation)
+- (Quickly) get resource information of this application in different clusters (cross-regional resource access)
+
+#### Story 2
+
+Imagine we have a resource management platform, which consists of multiple microservices, each of which needs to frequently acquire, display, and operate resources from different clusters
+
+Goals:
+
+- Provide a (cross-cluster) operation entry for users (view resources, events, logs, and operate resources)
```
Comment (RainbowMango) [inline docs/proposals/caching/README.md:48]: What kind of `operations` we are talking about? 
It makes sense for me that `read` resources from the cache would relieve the pressure of the member cluster, but the `update/delete` operation doesn't as we need to operate the clusters anyway.

File: docs/proposals/caching/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+Determine the target cluster and the corresponding resource type of the cache through the member cluster
+
+Implementation: Watch CRD, respond to ADD (cache cluster)/UPDATE (update cached object)/DELETE (remove cached cluster and cache information) events
+
+### API
+
+Aggregation API:
+
+- Compatible with Kubernetes: Authentication, Authorization(RBAC), Access Control
```
Comment (RainbowMango) [inline docs/proposals/caching/README.md:64]: Do you mean the `search` API which is installed on `karmada-apiserver` and users use this API to get resources from cache?

**Review by RainbowMango — COMMENTED** [review]
Hi sorry for letting this sit so long. 
We have just started the next release cycle, now it's time to discuss the proposal now.
We really want to have this feature in the next release(by the end of May).

**Comment by RainbowMango** [comment]
Note: I just updated the PR description(by adding a `NONE` to the release note part) to make the robot happy(removed `do-not-merge/release-note-label-needed`).

**Comment by XiShanYongYe-Chang** [comment]
Hi everybody, is there any progress on this proposal?

**Comment by huntsman-li** [comment]
I updated this proposal, modified the user story and design details

**Comment by RainbowMango** [comment]
Thanks @huntsman-li , I'm looking at it and will post my comments on it.

File: docs/proposals/caching/README.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+- Accelerate processing speed of resource requests across regions.
+- Get resources in multiple clusters by labels.
+
+## Design Details
+
+### Define the scope of the cached resource
+
+#### New ClusterCache API
+
+We propose a new CR in `cluster.karmada.io` group.
```
Comment (RainbowMango) [inline docs/proposals/caching/README.md:52]: Let's postpone making decisions for the group, as maybe we can introduce a new group, like `search.karmada.io`.

**Review by RainbowMango — COMMENTED** [review]
The review is still ongoing. @liys87x

File: docs/proposals/caching/README.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+	// Namespace of the target resource.
+	// Default is empty, which means all namespaces.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Indexes is a list of custom index
+	// Default is empty, which means no custom index.
+	// +optional
+	Indexes []ResourceIndexRef `json:"indexes,omitempty"`
```
Comment (RainbowMango) [inline docs/proposals/caching/README.md:137]: A question came to my mind: why user cares about the index thing? Does the cache( the informer cache) intend exposed to the user directly?

File: docs/proposals/caching/README.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    - kind: Pod
+    - kind: Ingress
+      apiVersion: networking.k8s.io/v1
+    - kind: DaemonSet
+      namespace: kube-system
+    - kind: Deployment
+      indexes:
+        - name: label
+          kinds:
+            - "*"
```
Comment (RainbowMango) [inline docs/proposals/caching/README.md:235]: The worlds `kind` and `Deployment` will be present twice if the index reference is restricted to Deployment.
It seems a little bit redundant and not elegant. Isn't?

```yaml
    - kind: Deployment
      indexes:
        - name: label
          kinds:
            - "Deployment"
```

**Review by RainbowMango — COMMENTED** [review]
Generally looks good to me.

I'll give more details suggestions later.

File: docs/proposals/caching/README.md (hunk 4 above)
Comment (liys87x) [reply docs/proposals/caching/README.md:137]: I agree, the ability provided by the cache does not need the user to care, I will remove index.

Comment (liys87x) [reply docs/proposals/caching/README.md:137]: What about remove it?

Comment (RainbowMango) [reply docs/proposals/caching/README.md:137]: +1
If the user doesn't care about it, then it should be removed from API.

Comment (liys87x) [reply docs/proposals/caching/README.md]: Removed it.

File: docs/proposals/caching/README.md (hunk 5 above)
Comment (liys87x) [reply docs/proposals/caching/README.md]: Removed it.

**Comment by RainbowMango** [comment]
I tried to tidy the commits but failed as this PR is sent from the `master` branch. 

@huntsman-li @liys87x I think we are OK on the API. And I'd like to merge this first and iterate it by following PRs.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

Let's iterate the proposal based on this.

**Pull request merged**